### DISCUSSION TITLE
Keycloak URL redirection issue when using non-standard ports with NGINX reverse proxy

### DISCUSSION THREAD

**Discussion by djkcyl** [keycloak/keycloak#33706] [Q&A]
**Description:**
I am using NGINX to reverse proxy Keycloak (version 26.0.0) on a server that cannot use standard HTTP and HTTPS ports (80 and 443). Instead, I am using ports 88 and 8443. My Keycloak instance is hosted on port 9610 and configured with `kc.http-port=9610`.

The reverse proxy works, but Keycloak keeps redirecting requests to `https://sso.xxx.com:443` instead of `https://sso.xxx.com:8443`, which is causing my setup to fail. Below are my configurations for both NGINX and Keycloak.

**Steps to reproduce:**
1. Set up NGINX to reverse proxy Keycloak with non-standard ports (88 for HTTP and 8443 for HTTPS).
2. Configure Keycloak to use port 9610 for HTTP.
3. Access Keycloak via the NGINX reverse proxy at `https://sso.xxx.com:8443`.
4. Observe that Keycloak redirects resources to `https://sso.xxx.com:443`, leading to failures.

**Expected behavior:**
Keycloak should honor the ports specified in the NGINX reverse proxy and redirect users to `https://sso.xxx.com:8443`, rather than defaulting to port 443.

**Actual behavior:**
Keycloak is redirecting requests and resources to `https://sso.xxx.com:443`, which causes connection issues because port 443 is not in use on my server.

**NGINX configuration:**
```nginx
# PROXY-START/
location ^~ /
{
    proxy_pass http://127.0.0.1:9610;
    proxy_set_header Host $host;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header X-Forwarded-Proto  $scheme;
    proxy_set_header REMOTE-HOST $remote_addr;
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection $connection_upgrade;
    proxy_http_version 1.1;

    add_header X-Cache $upstream_cache_status;

    set $static_fileOuhfr1Q5 0;
    if ( $uri ~* "\.(gif|png|jpg|css|js|woff|woff2)$" )
    {
        set $static_fileOuhfr1Q5 1;
        expires 1m;
    }
    if ( $static_fileOuhfr1Q5 = 0 )
    {
        add_header Cache-Control no-cache;
    }
}
# PROXY-END/
```

**Keycloak configuration (`kc.sh show-config`):**
```
Current Mode: production
Current Configuration:
        kc.db =  mysql
        kc.db-url =  jdbc:mysql://localhost/keycloak
        kc.db-username =  keycloak
        kc.hostname =  sso.xxx.com
        kc.http-enabled =  true
        kc.http-port =  9610
        kc.proxy-headers =  forwarded
        kc.spi-hostname-v2-hostname =  sso.xxx.com
        kc.version =  26.0.0
```

**Environment:**
- Keycloak version: 26.0.0
- NGINX version: [1.19.8]
- Database: MySQL

**Question:**
How can I configure Keycloak or NGINX to ensure that Keycloak redirects requests to the correct port (`8443`) instead of defaulting to `443` when accessed through the reverse proxy? Is there an additional configuration option or workaround I should be aware of to handle this situation?

**Thank you for your help!**

**Comment by djkcyl** [comment]
After further research, I updated the Keycloak hostname to https://sso.xxx.com:8443 and was able to access the admin console successfully.

New Issue:

However, I encountered a new problem. After logging in and clicking on "Manage Account," the page fails to load. In the F12 console, I see the following request returning a 403 error:

https://sso.xxx.com:8443/realms/master/protocol/openid-connect/login-status-iframe.html/init

Could you please advise on what might be causing this 403 error and how to resolve it?

Thank you!

**Reply by shawkins** [reply]
> Could you please advise on what might be causing this 403 error and how to resolve it?

You are likely failing a CORS check. You have kc.proxy-headers =  forwarded, but the proxy seems to be setup for xforwarded. Try that instead and see if it passes the check.

**Reply by djkcyl** [reply]
Thank you for your suggestion!

I updated my NGINX and Keycloak configurations as follows:

**NGINX:**
```nginx
# PROXY-START/
location ^~ /
{
    proxy_pass http://127.0.0.1:9610;
    proxy_set_header Host $host;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-Host sso.a60.one;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header X-Forwarded-Proto  $scheme;
}
# PROXY-END/
```

**Keycloak:**
```
kc.hostname = https://sso.a60.one:8443
kc.proxy-headers = xforwarded
```

Despite these changes, I am still encountering the same 403 error on the following request:

```
https://sso.a60.one:8443/realms/master/protocol/openid-connect/login-status-iframe.html/init
```

Any further advice on resolving this issue would be greatly appreciated!

**Reply by shawkins** [reply]
I believe X-Forwarded-Port must be included to satisfy the CORS check. You can turn on debug logging to see relevant logging https://github.com/keycloak/keycloak/blob/42484bd884a184e1078040bbacc450a578d07897/services/src/main/java/org/keycloak/protocol/oidc/endpoints/LoginStatusIframeEndpoint.java#L71 that should show why the request is being denied.

**Comment by djkcyl** [comment]
I added the `X-Forwarded-Port` header as suggested, and everything is now working perfectly. Thank you for your help!
